src/dex/swaap-v1/swaap-v1.ts

```diff
@@ -0,0 +1,633 @@
+    side: SwapSide,
+    state: SwaapV1PoolState,
+    currentTimestamp: bigint,
+  ): [bigint, bigint[]] {
+    if (side === SwapSide.SELL) {
+      let unitAmountOut: bigint;
+      try {
+        unitAmountOut = PoolQuotations.getAmountOutGivenInMMM(
+          state.liquidities[from],
+          state.oracles[from],
+          unitVolume,
+          state.liquidities[to],
+          state.oracles[to],
+          state.parameters,
+          currentTimestamp,
+        );
+      } catch (e) {
+        this.logger.error('Failed to get historical round data', e);
+        unitAmountOut = 0n;
+      }
+
+      const amountsOut: bigint[] = amounts.map(amountIn => {
+        try {
+          return PoolQuotations.getAmountOutGivenInMMM(
+            state.liquidities[from],
+            state.oracles[from],
+            amountIn,
+            state.liquidities[to],
+            state.oracles[to],
+            state.parameters,
+            currentTimestamp,
+          );
+        } catch (e) {
+          return 0n;
+        }
+      });
+
+      return [unitAmountOut, amountsOut];
+    } else {
+      let unitAmountIn;
+      try {
+        unitAmountIn = PoolQuotations.getAmountInGivenOutMMM(
+          state.liquidities[from],
+          state.oracles[from],
+          state.liquidities[to],
+          state.oracles[to],
+          unitVolume,
+          state.parameters,
+          currentTimestamp,
+        );
+      } catch (e) {
+        unitAmountIn = BigInt(MAX_UINT);
+      }
+
+      const amountsIn = amounts.map(amountOut => {
+        try {
+          return PoolQuotations.getAmountInGivenOutMMM(
+            state.liquidities[from],
+            state.oracles[from],
+            state.liquidities[to],
+            state.oracles[to],
+            amountOut,
+            state.parameters,
+            currentTimestamp,
+          );
+        } catch (e) {
+          return BigInt(MAX_UINT);
+        }
+      });
+      return [unitAmountIn, amountsIn];
+    }
+  }
+
+  // Returns pool prices for amounts.
+  // If limitPools is defined only pools in limitPools
+  // should be used. If limitPools is undefined then
+  // any pools can be used.
+  getPricesVolume(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    return this.getPricesVolumeLogic(
+      _from,
+      _to,
+      amounts,
+      side,
+      blockNumber,
+      null,
+      limitPools,
+    );
+  }
+
+  async getPricesVolumeLogic(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    _currentTimeStamp: bigint | null,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    const from = this.dexHelper.config.wrapETH(_from);
+    const to = this.dexHelper.config.wrapETH(_to);
+
+    const topPools = await this.getTopPoolsForTokens(
```

Comment:
`getTopPoolsForTokens` is doing http request to query the topPools. This is not good because the getPricesVolumes function is getting call a bunch of time on pricing request. You should be able to compute the topPools directly from memory

Reply:
We had a discussion regarding with the Paraswap team on TG .
To give you the context:
1. we have a dynamic list of pools, meaning we can’t know in advance what would be the pools
2. unfortunately we don’t have a factory method to retrieve this list
3. so we rely on subgraph to get this list (and to add new pools to the subscriber)
4. what would be your recommandation?
5. at the time there was 2 options:
a) to rely on the getTopPoolsForToken call to update ouf pools:  but in that case we would need this function to be called frequently (let’s say on a daily or weekly basis), to be sure that new pools are listed quickly on Paraswap.
b) to update the list of pools directly in the getPricesVolume: we chose this one because we understood that it was fine and that time (but maybe we misunderstood).

Would it be doable to stick to the option b. but with a condition on blockNumber so that we don't request TheGraph more than once by block?

Reply:
I think the solution **a** is the best solution, `getPricesVolume` is not made to handle this task. I think you can use a recursive `setTimeout` inside the `initializePricing` to do this task. 
Also new pool will probably have low liquidity so it will not impact the pricing that much.

Reply:
ok will do it